### PR TITLE
UCT/API: add error handling callback

### PR DESCRIPTION
**Pull request by evgeny-leksikov**
Should be merged before https://github.com/openucx/ucx/pull/1432

[UCP patch](https://github.com/evgeny-leksikov/ucx/commit/e01b5a873f2203c51ce5e5ac81542474fc6a5c13) based on this one in [development branch ](https://github.com/evgeny-leksikov/ucx/tree/ucp_err_handle_cb)

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1423/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3349/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1424/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3350/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
@yosefe pls, review

File: src/uct/api/uct.h [hunk 1]
```diff
@@ -1078,6 +1078,22 @@ ucs_status_t uct_iface_set_am_tracer(uct_iface_h iface, uct_am_tracer_t tracer,
...
+ * @brief Set error handler for the interface.
+ *
+ * The callback @a err_handler is called when transport level error is detected.
+ *
+ * @param [in]  iface           Interface to set the error handler for.
+ * @param [in]  error_handler   Callback to handle transport level error.
+ *                              NULL to clear.
+ * @param [in]  arg             Custom argument.
+ */
+ucs_status_t uct_iface_set_err_handler(uct_iface_h iface,
```
Comment (yosefe) [inline src/uct/api/uct.h]: the callback should be added to uct_iface_params_t

File: src/uct/api/uct_def.h [hunk 2]
```diff
@@ -223,6 +223,18 @@ typedef void (*uct_completion_callback_t)(uct_completion_t *self,
...
+
+/**
+ * @ingroup UCT_RESOURCE
+ * @brief Callback to process peer failure.
+ *
+ * @param [in]  ep       Endpoint to handle peer failure error,
+ *                       @a ep becomes unusable.
+ * @param [in]  arg      User argument to be passed to the callback.
+ */
+typedef void (*uct_error_handler_t)(uct_ep_h ep, void *arg);
```
Comment (yosefe) [inline src/uct/api/uct_def.h]: pls add ucs_status_t parameter

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1431/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3356/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1432/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3357/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3359/ for details (Mellanox internal link).

File: src/uct/base/uct_md.c [hunk 3]
```diff
@@ -360,14 +360,24 @@ ucs_status_t uct_iface_open(uct_md_h md, uct_worker_h worker,
...
     tlc = uct_find_tl_on_md(md->component, params->tl_name);
     if (tlc == NULL) {
         /* Non-existing transport */
         return UCS_ERR_NO_DEVICE;
     }
 
-    return tlc->iface_open(md, worker, params, config, iface_p);
+    status = tlc->iface_open(md, worker, params, config, iface_p);
+    if (status == UCS_OK) {
+        base_iface = ucs_derived_of((*iface_p), uct_base_iface_t);
```
Comment (yosefe) [inline src/uct/base/uct_md.c]: this code should be part of constructor of uct_base_iface_t

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1437/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3364/ for details (Mellanox internal link).

**Comment by abouteiller** [comment]
👍 +1

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1443/ for details.

**Comment by evgeny-leksikov** [comment]
@shamisp @bbenton pls, review as well

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1456/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3380/ for details (Mellanox internal link).

File: src/uct/api/uct_def.h [hunk 4]
```diff
@@ -223,6 +223,20 @@ typedef void (*uct_completion_callback_t)(uct_completion_t *self,
...
 
+
+/**
+ * @ingroup UCT_RESOURCE
+ * @brief Callback to process peer failure.
+ *
+ * @param [in]  arg      User argument to be passed to the callback.
+ * @param [in]  ep       Endpoint to handle peer failure error,
+ *                       @a ep becomes unusable. All following operation on @a ep
+ *                       will return @a status error.
```
Comment (bbenton) [inline src/uct/api/uct_def.h]: I suggest the following rewording of the description for the ep argument:
  Endpoint which has failed.  Upon return from the callback, this @a ep is no longer usable and all subsequent operations on this @a ep will fail with the error code passed in @a status

**Comment by shamisp** [comment]
What is the MellanoxLab failure ?

**Comment by shamisp** [comment]
@MattBBaker this change will break ugni

**Comment by yosefe** [comment]
@shamisp one of the VMs had network issue

**Comment by shamisp** [comment]
@yosefe okay. @MattBBaker will you be able to provide a patch.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3388/ for details (Mellanox internal link).

**Comment by MattBBaker** [comment]
@evgeny-leksikov I made a PR against your branch that adds UGNI support.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1470/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1471/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3392/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3393/ for details (Mellanox internal link).

**Comment by bbenton** [comment]
:+1: for Doc/API

**Comment by shamisp** [comment]
👍

**Pull request merged**